src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/SourceLauncher.java

```diff
@@ -201,6 +201,7 @@ private Class<?> execute(MemoryContext context, String[] mainArgs)
         try {
             ClassLoader loader = context.newClassLoaderFor(parentLoader, firstClassName);
             firstClass = Class.forName(firstClassName, false, loader);
+            Thread.currentThread().setContextClassLoader(loader);
```

Comment:
Hello Christian, a typical pattern for such Thread context classloader switches is to reset the context classloader back to the previous one, once the "work" is done. That way it doesn't influence the code after the completion of the work.

So something like the following might be more safer:

```java
ClassLoader loader;
try {
    loader = context.newClassLoaderFor(parentLoader, firstClassName);
    firstClass = Class.forName(firstClassName, false, loader);
} catch (ClassNotFoundException e) {
    throw new Fault(Errors.CantFindClass(firstClassName));
}
final ClassLoader prevTCCL = Thread.currentThread().getContextClassLoader();
Thread.currentThread().setContextClassLoader(loader);
try {
    Method mainMethod = MethodFinder.findMainMethod(firstClass);
    ... // rest of the existing code
    return mainClass;
} finally {
    Thread.currentThread().setContextClassLoader(prevTCCL);
}
```

Reply:
@jaikiran The PR is the equivalent of System.initPhase3 where it sets the TCCL for the main thread. With the source code launcher, the defining loader of the source "main" is a child of the system (usually application) class loader. So it's not meant to be temporary setting.

Reply:
Thank you Alan for that detail.